packages/block-library/src/index.js

```diff
@@ -317,39 +319,38 @@ export const registerCoreBlocks = (
 			registerBlockType( blockName, {
 				title: blockName,
 				edit: function Edit( { attributes } ) {
-					const { useBlockProps } = window.wp.blockEditor;
-					const { useServerSideRender } = window.wp.serverSideRender;
-					const { __, sprintf } = window.wp.i18n;
 					const blockProps = useBlockProps();
 					const { content, status, error } = useServerSideRender( {
 						block: blockName,
 						attributes,
 					} );
 
 					if ( status === 'loading' ) {
-						return createElement(
-							'div',
-							blockProps,
-							__( 'Loading…' )
+						return (
+							<div { ...blockProps }>{ __( 'Loading…' ) }</div>
 						);
 					}
 
 					if ( status === 'error' ) {
-						return createElement(
-							'div',
-							blockProps,
-							sprintf(
-								/* translators: %s: error message describing the problem */
-								__( 'Error loading block: %s' ),
-								error
-							)
+						return (
+							<div { ...blockProps }>
+								{ sprintf(
+									/* translators: %s: error message describing the problem */
+									__( 'Error loading block: %s' ),
+									error
+								) }
+							</div>
 						);
 					}
 
-					return createElement( 'div', {
-						...blockProps,
-						dangerouslySetInnerHTML: { __html: content || '' },
-					} );
+					return (
+						<div
```

Comment:
Are there auto-registered blocks who's wrapper is not a div?

Reply:
I don't see this as a RawHTML issue. The main issue here is that we need a React element to apply the props... The downside is that we only receive a string from the server that includes the full block. I think one potential solution is to not use `RawHTML` but something like `html-2-react` (I think libs like this exist) but I'm not sure about full compatibility. 

Also there are edge cases where the render callback returns more than a single html element top level. I think we'd need to handle these and consider these as not supported by auto-registered blocks.

Reply:
I understand that one of the reasons for not recommending the ServerSideRender package (or useServerSideRender hook) is that it creates the extra div wrapper in the Edit component. It would be great if we could make the DOM tree perfectly match on both the frontend and the editor. Do you guess this is technically possible?

Reply:
I think it's possible to try something like this yes, but I don't believe that's the reason for not using client side edit, client site edit is more about tailored UX.

Reply:
> client site edit is more about tailored UX.

I'm curious about this, for example, creating a React component dynamically based on a string defined on the PHP side?

Reply:
> I'm curious about this, for example, creating a React component dynamically based on a string defined on the PHP side?

No I meant that we don't recommend ServerSideRender because the ideal experience is that the block in the editor is interactive and updates as you make block modifications without reaching the server, which is not the case for ServerSiteRender.

---

That said, that doesn't stop us from making ServerSideRender better for those who use it and server side registered block. And using a library like you shared could be a good path forward to make the editor markup match the frontend.